### DISCUSSION TITLE
How to use `H5Version.hpp.in` to access `HIGHFIVE_VERSION_*` definitions

### DISCUSSION THREAD

**Discussion by Schellenberg3** [BlueBrain/HighFive#778] [Questions] [answered]
I've cloned HighFive and have been using it with my project. For my project, it would be nice to track group creation order but it is not required. Since this feature was added in 2.7.0 I would like something like this:

```cpp
HighFive::GroupCreateProps properties{};

#if HIGHFIVE_VERSION_MAJOR > 2 && HIGHFIVE_VERSION_MINOR > 7
properties.add(HighFive::LinkCreationOrder(HighFive::CreationOrder::Tracked | HighFive::CreationOrder::Indexed));
#endif

auto sensor_record_group = metrics_group.createGroup("SensorRecord", properties);
```

So we only try to compile the features if the correct version is available.

I see that these definitions exist in the file [`H5Version.hpp.in`](https://github.com/BlueBrain/HighFive/blob/be68bd0efcef338a016fba448d6444089fd196d5/include/highfive/H5Version.hpp.in#L11-L13) but I am unfamiliar with how the file is formatted:

https://github.com/BlueBrain/HighFive/blob/be68bd0efcef338a016fba448d6444089fd196d5/include/highfive/H5Version.hpp.in#L11-L13

From this [Stack Overflow question](https://stackoverflow.com/questions/14248848/) it seems like the files `*.in` extension and `@...@` formatting is related to [autoconf](http://www.gnu.org/software/autoconf/manual/autoconf.html).

Could anyone provide some guidance on how to properly use this file?

For context, I have HighFive as a submodule for my project and am developing on a Windows machine.

**Comment by 1uc** [comment]
The `*.in` file and `@...@` templating syntax is related to CMake. As a user of HighFive `H5Version.hpp.in` is interesting because something (cmake) will create a file `H5Version.hpp`. Essentially, user code should do:

```
#include <highfive/H5Version.hpp>

// ...

HighFive::GroupCreateProps properties{};

#if HIGHFIVE_VERSION_MAJOR > 2 && HIGHFIVE_VERSION_MINOR > 7
properties.add(HighFive::LinkCreationOrder(HighFive::CreationOrder::Tracked | HighFive::CreationOrder::Indexed));
#endif

auto sensor_record_group = metrics_group.createGroup("SensorRecord", properties);
```

I'd suggest you first try if you're lucky and just include the above file and use the macros `HIGHFIVE_VERSION_{MAJOR,MINOR,PATCH}`. 

If that doesn't work the most likely reason is that the file hasn't been generated (e.g. if your project doesn't use CMake or doesn't have a `add_subdirectory(highfive)`). Alternatively, it might have been generated, but because it's generated in the build directory of HighFive a `-I${HIGHFIVE_BUILD_DIR}` might be missing (e.g. because your project doesn't use the HighFive CMake target). If you're in this case, please describe in a little bit more detail how you're using HighFive. I think the important questions are:

1. Do you use CMake?
2. If yes, do you use `add_subdirectory` and or the CMake target `HighFive`?
3. If no, do you install HighFive (`cmake -D ... -B build . && cmake --build build && cmake --install build`) or use the sources directly?

**Reply by Schellenberg3** [reply]
Thanks for the response, Luc!

I am using CMake in my project. From my file, the lines relevant to HighFive are:

```cmake
set(HIGHFIVE_UNIT_TESTS OFF CACHE BOOL "Enable unit tests(requires Catch2 to be present)")
set(HIGHFIVE_BUILD_DOCS OFF CACHE BOOL "Enable documentation building")
set(HIGHFIVE_EXAMPLES   OFF CACHE BOOL "Compile examples")

add_subdirectory(${BASE_DIR}/deps/HighFive)
set(HIGHFIVE_INCLUDE_DIRS ${PROJECT_SOURCE_DIR}/deps/HighFive/include)

# Other packages imports with find_package...

include_directories(
    # Other includes ...
    ${HIGHFIVE_INCLUDE_DIRS}
)
```

CMake is still new to me so I would not be surprised if I am making a simple mistake!

Looking at the `CMakeLists.txt` in HighFive's root I see where the line where CMake should be generating this file:

https://github.com/BlueBrain/HighFive/blob/be68bd0efcef338a016fba448d6444089fd196d5/CMakeLists.txt#L10-L11

However, when I build my project, the `*.hpp.in` file is not turned into the expected `*.hpp` file.

I don't need the unit tests, docs, or examples but could setting those flags off have impacted something?

**Reply by 1uc** [reply, marked as answer]
The first three lines are good, you don't need any of that. (You might want to also remove Boost support; we can do that later.)

The line
```
add_subdirectory(${BASE_DIR}/deps/HighFive)
```
should trigger the `configure_file`. (Unrelated: check the `BASE_DIR` it should probably be some `CMAKE_*_DIR`.) The `add_subdirectory` should also make the target `HighFive` available. You'd use it as follows
```
add_executable(runme ...)
target_link_libraries(runme PUBLIC HighFive)

add_library(linkme ...)
target_link_libraries(linkme [PRIVATE,PUBLIC] HighFive)
```

Please remove `${HIGHFIVE_INCLUDE_DIRS}` from:
```
include_directories(
    # Other includes ...
    ${HIGHFIVE_INCLUDE_DIRS}
)
```
Generally, this isn't modern cmake; and the "modern" is what makes cmake much more agreeable. Personally, I like this reference (and references therein):

https://cliutils.gitlab.io/modern-cmake/chapters/basics.html

Now, you should check the compile commands:
```
cmake --build build --verbose
```
It should include something vaguely like:
```
/usr/bin/c++ ... -I${YOUR_INCLUDE_PATHS} -I deps/HighFive/include -I build/.../deps/HighFive/include -isystem /usr/include/hdf5/serial -O2 -g -DNDEBUG -std=c++11 -fPIC  ...
```
The salient point being that there's two include paths for HighFive, one is simply the directory where the code resides, the other is a subdirectory of the build directory (which is where you should find `H5Version.hpp`).

**Reply by Schellenberg3** [reply]
This was exactly the issue! It seems there was something with my misuse of CMake.

I updated my project's `CMakeLists.txt` to utilize `target_link_libraries`. I also turned off boost, as you suggested, and make some other, improvements suggested in the introduction of _Modern CMake_. Somewhere along the way, the `configure_file` line was properly activated.

Running the `cmake --build build --verbose` command I did see the two include paths you highlighted and found the `H5Version.hpp` file within the build directory. My code is including this correctly now!

Thank you for the assistance and for providing a great resource for learning about CMake!
